File: src/ast.jl
# MinRust AST - Abstract Syntax Tree Definitions

module MinRustAST

export ASTNode, Program, FnDecl, VarDecl, StructDecl, ImplBlock,
       Block, Param, StructField,
       Expr, Literal, Identifier, BinaryOp, UnaryOp, Call, MethodCall,
       FieldAccess, Index, ArrayLiteral, VecMacro, StructLiteral, Cast, Range,
       Stmt, IfStmt, ForStmt, WhileStmt, LoopStmt, MatchStmt, ReturnStmt, ExprStmt,
       RustType, PrimitiveType, ReferenceType, ArrayType, StructType

# ============================================================
# Base Types
# ============================================================

abstract type ASTNode end
abstract type Expr <: ASTNode end
abstract type Stmt <: ASTNode end
abstract type RustType end

# ============================================================
# Top-level Program
# ============================================================

struct Program <: ASTNode
    items::Vector{ASTNode}
end

# ============================================================
# Type Definitions
# ============================================================

struct PrimitiveType <: RustType
    name::String  # "i32", "i64", "f64", "bool", "String", etc.
end

struct ReferenceType <: RustType
    inner::RustType
    is_mutable::Bool
end

struct ArrayType <: RustType
    element_type::RustType
    size::Union{Int, Nothing}  # Nothing for slices
end

struct StructType <: RustType
    name::String
end

struct VecType <: RustType
    element_type::RustType
end

struct FunctionType <: RustType
    param_types::Vector{RustType}
    return_type::RustType
end

# ============================================================
# Function and Struct Declarations
# ============================================================

struct Param
    name::String
    type_annotation::RustType
end

struct FnDecl <: ASTNode
    name::String
    params::Vector{Param}
    return_type::Union{RustType, Nothing}
    body::Block
end

struct StructField
    name::String
    type_annotation::RustType
end

struct StructDecl <: ASTNode
    name::String
    fields::Vector{StructField}
end

struct ImplBlock <: ASTNode
    for_type::String
    methods::Vector{FnDecl}
end

# ============================================================
# Statements
# ============================================================

struct Block <: ASTNode
    statements::Vector{ASTNode}
end

struct VarDecl <: Stmt
    name::String
    type_annotation::Union{RustType, Nothing}
    is_mutable::Bool
    init::Union{Expr, Nothing}
end

struct ExprStmt <: Stmt
    expr::Expr
end

struct IfStmt <: Stmt
    condition::Expr
    then_block::Block
    else_block::Union{Block, Nothing}
end

struct ForStmt <: Stmt
    var::String
    iter_expr::Expr
    body::Block
end

struct WhileStmt <: Stmt
    condition::Expr
    body::Block
end

struct LoopStmt <: Stmt
    body::Block
end

struct MatchStmt <: Stmt
    expr::Expr
    arms::Vector{MatchArm}
end

struct MatchArm
    pattern::String  # simplified: just pattern string
    expr::Expr
end

struct ReturnStmt <: Stmt
    value::Union{Expr, Nothing}
end

struct BreakStmt <: Stmt
    label::Union{String, Nothing}
end

struct ContinueStmt <: Stmt
    label::Union{String, Nothing}
end

# ============================================================
# Expressions
# ============================================================

struct Literal <: Expr
    value::Union{Int, Float, String, Char, Bool}
end

struct Identifier <: Expr
    name::String
end

struct BinaryOp <: Expr
    left::Expr
    op::String  # "+", "-", "*", "/", "==", "!=", "<", ">", "<=", ">=", "&&", "||", etc.
    right::Expr
end

struct UnaryOp <: Expr
    op::String  # "-", "!", "&", "*", "~"
    operand::Expr
end

struct Call <: Expr
    func::Union{String, Expr}  # function name or expression
    args::Vector{Expr}
end

struct MethodCall <: Expr
    object::Expr
    method::String
    args::Vector{Expr}
end

struct FieldAccess <: Expr
    object::Expr
    field::String
end

struct Index <: Expr
    array::Expr
    index::Expr
end

struct ArrayLiteral <: Expr
    elements::Vector{Expr}
end

struct VecMacro <: Expr
    elements::Vector{Expr}
end

struct StructLiteral <: Expr
    struct_name::String
    fields::Vector{Pair{String, Expr}}
end

struct Cast <: Expr
    expr::Expr
    target_type::RustType
end

struct Range <: Expr
    start::Union{Expr, Nothing}
    end::Union{Expr, Nothing}
    inclusive::Bool
end

struct IfExpr <: Expr
    condition::Expr
    then_block::Block
    else_block::Union{Block, Nothing}
end

struct MatchExpr <: Expr
    expr::Expr
    arms::Vector{MatchArm}
end

struct Closure <: Expr
    params::Vector{String}
    body::Block
end

struct StringFormatExpr <: Expr
    format_str::String
    args::Vector{Expr}
end

end  # module MinRustAST

File: src_rust/src/ast.rs
// MinRust AST - Phase 5 Rust Implementation
// Abstract Syntax Tree structures

use crate::tokenizer::Token;

/// Top-level program structure
#[derive(Debug, Clone)]
pub struct Program {
    pub items: Vec<Item>,
}

/// Top-level items
#[derive(Debug, Clone)]
pub enum Item {
    Function(FnDecl),
    Struct(StructDecl),
    ImplBlock(ImplBlock),
}

/// Function declaration
#[derive(Debug, Clone)]
pub struct FnDecl {
    pub name: String,
    pub params: Vec<Param>,
    pub return_type: Option<RustType>,
    pub body: Vec<Stmt>,
}

/// Function parameter
#[derive(Debug, Clone)]
pub struct Param {
    pub name: String,
    pub ty: RustType,
}

/// Struct declaration
#[derive(Debug, Clone)]
pub struct StructDecl {
    pub name: String,
    pub fields: Vec<StructField>,
}

/// Struct field
#[derive(Debug, Clone)]
pub struct StructField {
    pub name: String,
    pub ty: RustType,
}

/// Implementation block
#[derive(Debug, Clone)]
pub struct ImplBlock {
    pub target: String,
    pub methods: Vec<FnDecl>,
}

/// Statements
#[derive(Debug, Clone)]
pub enum Stmt {
    VarDecl {
        name: String,
        ty: Option<RustType>,
        init: Option<Expr>,
        mutable: bool,
    },
    ConstDecl {
        name: String,
        ty: Option<RustType>,
        value: Expr,
    },
    IfStmt {
        condition: Expr,
        then_body: Vec<Stmt>,
        else_body: Option<Vec<Stmt>>,
    },
    ForStmt {
        var: String,
        iter: Expr,
        body: Vec<Stmt>,
    },
    WhileStmt {
        condition: Expr,
        body: Vec<Stmt>,
    },
    LoopStmt {
        body: Vec<Stmt>,
    },
    MatchStmt {
        expr: Expr,
        arms: Vec<(String, Vec<Stmt>)>,
    },
    ReturnStmt(Option<Expr>),
    ExprStmt(Expr),
    Block(Vec<Stmt>),
}

/// Expressions
#[derive(Debug, Clone)]
pub enum Expr {
    Literal(Literal),
    Identifier(String),
    BinaryOp {
        left: Box<Expr>,
        op: String,
        right: Box<Expr>,
    },
    UnaryOp {
        op: String,
        operand: Box<Expr>,
    },
    Call {
        func: String,
        args: Vec<Expr>,
    },
    MethodCall {
        object: Box<Expr>,
        method: String,
        args: Vec<Expr>,
    },
    FieldAccess {
        object: Box<Expr>,
        field: String,
    },
    Index {
        object: Box<Expr>,
        index: Box<Expr>,
    },
    ArrayLiteral(Vec<Expr>),
    StructLiteral {
        name: String,
        fields: Vec<(String, Expr)>,
    },
    Cast {
        expr: Box<Expr>,
        ty: RustType,
    },
    Range {
        start: Option<Box<Expr>>,
        end: Option<Box<Expr>>,
    },
}

/// Literal values
#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    Integer(i64),
    Float(f64),
    String(String),
    Char(char),
    Bool(bool),
}

/// Rust types
#[derive(Debug, Clone, PartialEq)]
pub enum RustType {
    Primitive(String), // i32, f64, bool, char, String, etc
    Reference {
        inner: Box<RustType>,
        mutable: bool,
    },
    Array {
        element: Box<RustType>,
        size: Option<usize>,
    },
    Vec(Box<RustType>),
    Struct(String),
    Function {
        params: Vec<RustType>,
        return_ty: Box<RustType>,
    },
}

impl RustType {
    /// Check if type is primitive
    pub fn is_primitive(&self) -> bool {
        matches!(self, RustType::Primitive(_))
    }

    /// Check if type is numeric
    pub fn is_numeric(&self) -> bool {
        match self {
            RustType::Primitive(name) => {
                matches!(name.as_str(), "i32" | "i64" | "f64" | "u32" | "u64")
            }
            _ => false,
        }
    }

    /// Get string representation
    pub fn to_string_repr(&self) -> String {
        match self {
            RustType::Primitive(name) => name.clone(),
            RustType::Reference { inner, mutable } => {
                let mut_str = if *mutable { "mut " } else { "" };
                format!("&{}{}", mut_str, inner.to_string_repr())
            }
            RustType::Array { element, size } => {
                if let Some(s) = size {
                    format!("[{}; {}]", element.to_string_repr(), s)
                } else {
                    format!("[{}]", element.to_string_repr())
                }
            }
            RustType::Vec(inner) => format!("Vec<{}>", inner.to_string_repr()),
            RustType::Struct(name) => name.clone(),
            RustType::Function { params, return_ty } => {
                let param_str = params
                    .iter()
                    .map(|p| p.to_string_repr())
                    .collect::<Vec<_>>()
                    .join(", ");
                format!("fn({}) -> {}", param_str, return_ty.to_string_repr())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_type_primitive() {
        let ty = RustType::Primitive("i32".to_string());
        assert!(ty.is_primitive());
        assert!(ty.is_numeric());
    }

    #[test]
    fn test_type_reference() {
        let inner = Box::new(RustType::Primitive("i32".to_string()));
        let ty = RustType::Reference {
            inner,
            mutable: false,
        };
        assert!(!ty.is_primitive());
    }
}
